identity/registry/registry_contract.go

```diff
@@ -204,109 +218,78 @@ func (registry *contractRegistry) handleRegistrationEvent(ev IdentityRegistratio
+			log.Info().Msg("registration watch subscription timed out")
+			registry.saveRegistrationStatus(ev.ChainID, ev.Identity, RegistrationError)
 			return
-		case <-sink:
-			log.Info().Msgf("Received registration event for %v", identity)
-			_, err := registry.storage.Get(chainID, identity)
+		case <-time.After(registry.cfg.TransactorPollInterval):
+			res, err := registry.transactor.FetchRegistrationStatus(ev.Identity)
 			if err != nil {
-				log.Error().Err(err).Msg("Could not store registration status")
+				log.Warn().Err(err).Msg("could not fetch registration status from transactor")
+				break
 			}
 
-			status := Registered
-
-			log.Debug().Msgf("Sending registration success event for %v", identity)
-			registry.publisher.Publish(AppTopicIdentityRegistration, AppEventIdentityRegistration{
-				ID:      identity,
-				Status:  status,
-				ChainID: chainID,
-			})
-
-			err = registry.storage.Store(StoredRegistrationStatus{
-				Identity:           identity,
-				RegistrationStatus: status,
-				ChainID:            chainID,
-			})
-			if err != nil {
-				log.Error().Err(err).Msg("Could not store registration status")
+			var resp *TransactorStatusResponse
+			for _, v := range res {
+				if v.ChainID != ev.ChainID {
+					continue
+				}
+				resp = &v
+				break
 			}
-		case err := <-subscription.Err():
-			if err == nil {
-				return
+
+			if resp == nil {
+				log.Warn().Msg("no matching registration entries for chain in transactor response, will continue")
+				break
 			}
 
-			log.Error().Err(err).Msg("Subscription error")
-			registry.publisher.Publish(AppTopicIdentityRegistration, AppEventIdentityRegistration{
-				ID:      identity,
-				Status:  RegistrationError,
-				ChainID: chainID,
-			})
-			updateErr := registry.storage.Store(StoredRegistrationStatus{
-				Identity:           identity,
-				RegistrationStatus: RegistrationError,
-				ChainID:            chainID,
-			})
-			if updateErr != nil {
-				log.Error().Err(updateErr).Msg("could not store registration status")
+			switch resp.Status {
+			case TransactorRegistrationEntryStatusSucceed:
+				registry.resyncWithBC(ev.ChainID, ev.Identity)
+			case TransactorRegistrationEntryStatusFailed:
+				log.Error().Msg("registration reported as failed by transactor")
+				registry.saveRegistrationStatus(ev.ChainID, ev.Identity, RegistrationError)
+				return
 			}
 		}
-	}()
+	}
+}
+
+func (registry *contractRegistry) resyncWithBC(chainID int64, id string) {
+	status, err := registry.bcRegistrationStatus(chainID, identity.FromAddress(id))
+	if err != nil {
+		log.Err(err).Msg("could not check registration status on chain")
+		registry.saveRegistrationStatus(chainID, id, RegistrationError)
+		return
+	}
+	registry.saveRegistrationStatus(chainID, id, status)
+}
+
+func (registry *contractRegistry) saveRegistrationStatus(chainID int64, id string, status RegistrationStatus) {
+	registry.publisher.Publish(AppTopicIdentityRegistration, AppEventIdentityRegistration{
```

Comment:
maybe it would be better if we stored first then published? Then we'd know that what we published is the same thing we've stored.